### PR TITLE
Fix invalid suggestion of changing impl trait signature

### LINKED ISSUE #68049: Invalid suggestion to change `&self` to `&mut self` in trait implementation
**Summary**: The compiler suggests changing `&self` to `&mut self` when trying to modify `self` in a method. For trait implemantions this does not work because it would require changing the trait definition too (which is often not possible).

**Example**: 

I was trying to implement the `GlobalAllocator` trait from the standard library for a custom type:

```rust
use std::alloc::{GlobalAlloc, Layout};

struct Test(u32);

unsafe impl GlobalAlloc for Test {
    unsafe fn alloc(&self, _layout: Layout) -> *mut u8 {
        self.0 += 1;
        0 as *mut u8
    }

    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {
        unimplemented!();
    }
}
```

([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=519036c88db0aeb46748aa4640d2453a))

This of course doesn't work since you can't modify data behind a `&self` reference:

```
error[E0594]: cannot assign to `self.0` which is behind a `&` reference
 --> src/main.rs:7:9
  |
6 |     unsafe fn alloc(&self, _layout: Layout) -> *mut u8 {
  |                     ----- help: consider changing this to be a mutable reference: `&mut self`
7 |         self.0 += 1;
  |         ^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be written
```

However, I noticed that the _help_ message suggests changing the `&self` to `&mut self`, which is incorrect since that would require changing the trait definition in the standard library too.

**Suggestion:** I don't know if this is possible, but maybe this help message should only be shown for "normal" methods and hidden when implementing trait methods?

### PR DESCRIPTION
**Pull request by hkalbasi**
Fix #68049

### CODE REVIEW

**Comment by HKalbasi** [comment]
Seems @matthewjasper is busy (from github contributions). May I request another reviewer?
r? @Aaron1011

File: compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs [hunk 1]
```diff
@@ -424,15 +424,108 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+                                /// out of scope of this crate
+                                let (is_trait_sig, local_trait) = {
+                                    if self.body.local_kind(local) != LocalKind::Arg {
+                                        (false, None)
+                                    } else {
+                                        let hir_map = self.infcx.tcx.hir();
+                                        let my_hir = hir_map.local_def_id_to_hir_id(
+                                            self.body.source.def_id().as_local().unwrap(),
+                                        );
+                                        match hir_map.find(hir_map.get_parent_node(my_hir)) {
```
Comment (Aaron1011) [inline compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs:440]: You can use `tcx.opt_associated_item` to determine if the method comes from a trait impl.

**Comment by Aaron1011** [comment]
Thank you for your contribution!

There's a large amount of rightward drift in the diff. If it's still there after you switch to `tcx.opt_associated_item`, could you make a new method for the logic you're adding? That will allow you to return early when a condition doesn't match, which should help keep the rightward drift down.

File: compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs (hunk 1 above)
Comment (HKalbasi) [reply compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs]: I also used `my_hir` in `hir_map.name(hi) != hir_map.name(my_hir)` as well. If problem is count of characters, I don't think `tcx.opt_associated_item` help much?

**Comment by HKalbasi** [comment]
I moved the logic to a function so right drift is almost solved. I don't think it helps to use `tcx.opt_associated_item` because I need trait object (`td`) and `my_hir`.

@rustbot label -S-waiting-on-author +S-waiting-on-review

File: compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs [hunk 2]
```diff
@@ -502,6 +515,84 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
+    /// suggestion to the trait or disable it if it is out of scope of this crate                            
+    fn is_error_in_trait(&self, local: Local) -> (bool, Option<Span>) {
+        if self.body.local_kind(local) != LocalKind::Arg {
+            return (false, None);
+        }
+        let hir_map = self.infcx.tcx.hir();
+        let my_hir = hir_map.local_def_id_to_hir_id(
+            self.body.source.def_id().as_local().unwrap(),
+        );
+        match hir_map.find(hir_map.get_parent_node(my_hir)) {
```
Comment (Aaron1011) [inline compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs:530]: This long match can be replaced by `tcx.opt_associated_item` to get an `AssocItem`, and then `tcx.trait_id_of_impl` to get the `DefId` of the trait.

**Comment by Aaron1011** [comment]
I see what you mean about still needing to use the HIR to get the trait method argument span. However, using `opt_associated_item` should make some of the code preceding that loop much shorter

File: compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs (hunk 2 above)
Comment (HKalbasi) [reply compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs]: I replaced long match with `tcx.impl_of_method`.

**Comment by HKalbasi** [comment]
@Aaron1011 Is it now ok?

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/85163#issuecomment-837520195
Note that someone should approve the workflow for first-time contributors.
@bors r-

**Comment by HKalbasi** [comment]
I thought `./x.py test tidy --bless` will fix styling problems automatically.

**Comment by HKalbasi** [comment]
What is happening here? `tidy --bless` fix styling but just have problem with trailing whitespaces?

**Comment by HKalbasi** [comment]
But trailing space is the easiest one I think. Maybe it is a bug?

**Comment by JohnTitor** [comment]
> But trailing space is the easiest one I think. Maybe it is a bug?

Nah, it isn't a _bug_, `--bless` means `x.py fmt` and it won't remove trailing spaces. Just checked on my local and tidy exits with code 1 correctly so I think you missed the failure on your console.

**Comment by JohnTitor** [comment]
Could you squash extra formatting commits into the existing one to clean up the commit log? Once it's done, I'll back this into the queue.

**Comment by HKalbasi** [comment]
Done

**Pull request merged**